### PR TITLE
DG GR Maxwell (Vacuum)

### PR DESCRIPTION
**Pull request by johnson452**
# Feature
The General Relativistic Maxwell’s Equations system.

## Summary
As a first step towards the coupled DG-GR-Maxwell-Vlasov system, I’ve implemented the DG-GR Maxwell equations in vacuum. These are solved using a Lax flux or Roe flux with hard-coded for Roe to spherical Kerr-Schild (Roe only), in the GR variables. See component forms of the equations. Note that no coupling has been done as of yet and geometry is assumed stationary.

## Implementation Details

Key changes: _List the key changes made to the codebase to implement the feature._

1.	The surface terms are handles nodally where geometry is evaluated exactly and fields are evaluated at nodal points to comptue the surface fluxes.
2.	Volume nodes are also computed similarly.
3.	Geomtry information is assumed stationary and pre-computed when calling DG-GR modules. This includes in Maxwell: (alpha, beta contravariant, h_ij covariant, and J_c) and for kinetics: the Hamiltonian and poisson tensor.
4.	Vlasov geometry now uses a structure: gkyl_surf_and_vol_nodes_arrays which stores (component major) arrays of nodes at volume gauss-legandre points in nodal_arr_vol, and surface nodal arrays in x, y, z: nodal_arr_surf_{x,y,z}.
5.	Evolved variables include jacobian factors: Q = [ Jc D^i, Jc B^i, Jc phi, Jc psi ]. Division routines like Maxwell can remove or add J_c to the fields, similar to Vlasov functions, but here these are over configruations space only and divide out just the field componets.
6.	See code examples for API changes, such as the field model id type (GKYL_FIELD_GR_D_B), and theta boundary condition (bcThetaPole) in a lua script example.
7.	Polyorder one and two are both supported.

Supported geometries: (continuously updating with more in Vlasov_lw.c and specified in preset geometry fields for kinetics and maxwells). The format here is Lua/C:
  { "None", GKYL_TRIAD_NONE },
  { "Flat", GKYL_TRIAD_FLAT },
  { "Annulus", GKYL_TRIAD_ANNULUS },
  { "Cylindrical_rz", GKYL_TRIAD_CYLINDRICAL_RZ },
  { "Spherical_rtheta", GKYL_TRIAD_SPHERICAL_RTHETA },
  { "GR_KS_rphi", GKYL_TRIAD_GR_KERR_SCHILD_RPHI },
  { "GR_KS_r", GKYL_TRIAD_GR_KERR_SCHILD_R },
  { "GR_KS_rtheta", GKYL_TRIAD_GR_KERR_SCHILD_RTHETA },
  { "GR_KS_3V", GKYL_TRIAD_GR_KERR_SCHILD_3V },
  { "GR_KS_Cart_3V", GKYL_TRIAD_CART_GR_KERR_SCHILD_3V },

Automated testing: _Please explain how the feature was tested. If no automated tests are included (e.g. unit, regression), explain why._
Unit tests:
-	(ctest_dg_gr_maxwell_geom.c) Test of 
o	Field initizalization.
o	Geometry initialization at nodal points on surfaces and volumes (alpha, beta^i, h_ij, J_c.
o	Division by J_c.
o	Boundary conditions, including a new theta-pole boundary condition for spherical coordiantes with fields.
o	Configuration space flux and maximum eigenvalue calculation.
o	Flux jump condition testing.
Regression tests:
-	rt_dg_gr_maxwell_1d_riemann.lua – Lua test of a 1D jump in fields from (AH: https://ammar-hakim.org/sj/je/je6/je6-maxwell-solvers.html#problem-3-1d-riemann-problem) 
-	rt_dg_gr_maxwell_cylindrical_wg.lua (and .c equivalent) – Lua side test of a standing cyclindrical waveguide mode. (Matches AH configuration: https://ammar-hakim.org/sj/je/je36/je36-maxwell-cylindrical-waveguide.html) C side test was created for sanitzer testings.
-	rt_dg_gr_maxwell_rectangular_wg.lua – Rectangular waveguide case with dg-gr-maxwell infrastructure. Matches the rt_dg_maxwell testcase parameters.
-	rt_dg_gr_maxwell_spherical_wg.lua – Testing a spherical cavity mode and theta-pole BCs with an azimuthally symmetric mode in the r-theta plane.
-	rt_dg_gr_maxwell_wald.lua – See S. S. Komissarov’s 2004 magnetospheres paper for reference. Tests the spin free horizon-penetrating coordinate stabiltiy of the Wald solution.
-	rt_dg_gr_maxwell_wald_kerr.lua –Tests the the spinning horizon-penetrating coordinate stabiltiy of the Wald solution.

## Example Use

_Provide some example code of how a user may utilize this new feature in an input file._
A typical (lua) input file requires:
1.	Geometry needs to be specificed
```
vlasovApp = Vlasov.App.new {

  -- Vlasov Geometry
  geom = Vlasov.Geom.new {
    
    -- Black hole parameters
    massBH = massBH,
    spinBH = spinBH,

    -- Specify the geometry rtheta choice
    usePresetGeom = true,
    triadPresetGeomType = G0.TriadGeom.GR_KS_rtheta
    
  },
…
```
2.	D and B fields, in contravariant componenets must be specified for inputs

```
-- Field.
  field = Vlasov.Field.new {

    -- Use GR field ID
    fieldID = G0.FieldModel.GR,

    -- Initial conditions function.
    init = function (t, xn)
      local r, theta = xn[1], xn[2]

      -- D^i, and B^i (contravaraint components)
      -- Even though E_i is non-zero, the vacuum consituitve relations give 
      -- a non-zero D^\phi component
      local Dr = 0.0 -- Total electric field (r-direction).
      local Dtheta = 0.0 -- Total electric field (theta-direction).
      local Dphi = - 2.0 * massBH * B_0 / ( r * r * math.sqrt( 1.0 + 2.0 * massBH / r ) )  -- Total electric field (phi-direction).
    
      local Br = - B_0 * math.cos(theta) / ( math.sqrt( 1.0 + 2.0 * massBH / r ) )   -- Total magnetic field (r-direction).
      local Btheta = B_0 * math.sin(theta) / ( r * math.sqrt( 1.0 + 2.0 * massBH / r ) ) -- Total magnetic field (theta-direction).
      local Bphi = 0.0 -- Total magnetic field (phi-direction).

      -- Hand off the primative variables (U^\xi)
      return Dr, Dtheta, Dphi, Br, Btheta, Bphi, 0.0, 0.0
    end,

    -- Copy boundary conditions in r are sufficient. Theta requries theta-pole BCs
    bcx = { G0.FieldBc.bcFixedFunc, G0.FieldBc.bcFixedFunc }, -- boundary conditions (r-direction).
    bcy = { G0.FieldBc.bcThetaPole, G0.FieldBc.bcThetaPole } -- boundary conditions (theta-direction).
  }

```

## Community Standards

- [ ] Documentation has been updated.
- [ ] My code follows the project's coding guidelines.
- [x] Changes to `layer/zero` should have a unit test, e.g., `core/zero`.

## Testing: _(x (yes), blank (no))_

- [x] I added a regression test to test this feature.
- [ ] I added this feature to an existing regression test.
- [x] I added a unit test to test this feature.
- [x] Ran `make check` and unit tests all pass.
- [x] I ran the code with make valcheck, and it is clean.
- [x] I ran the code through computer-sanitizer on GPU, and it is clean.
- [x] I ran a few regression tests to ensure no apparent errors.
- [x] Tested and works on CPU.
- [ ] Tested and works on multi-CPU.
- [x] Tested and works on GPU.
- [x] Tested and works on multi-GPU.

## Additional Notes

_Include any additional context, caveats, or future improvements related to the feature._
-	Natural units (c = eps0 = mu0 = 1) are assumed everywhere. A check is inplace if someone tries for something other than c = 1 with GR maxwells.
-	Field_id is primarily utilized to determine program flow differences between default (shared) behavior and gr-maxwell’s.
-	h_ij is always assumed to be an upper triangular metric of 6 unique elements 
-	use_lo will only be for kinetics (removed from conf-kernels for dg_gr_maxwell since there is no benefit here.)
-	Copies exactly the structure for static jacobian removal from JJ’s work. The Jaocbian is first removed before computing and surface terms, then Jc is exact at required location.
-	Dg maxwell uses an inew method now

FUTURE INVESTIGATIONS:

-	Seems that Vlasov triad can be refactored to do spatial geometry like this to include the theta-pole boundary for full axisymmetric problems. Since phi isn't resolved, the CFL by approaching theta = 0 isn't exploding by moving across small phi cells
-	PHM is left to a separate branch and commit since it will heavily weave through everything in a non-trivial way.

FIGURES:
1. The basic equations being solved:
image
2. The surface node and surface volume geometry:
image

### CODE REVIEW

File: vlasov/apps/vm_species.c [hunk 1]
```diff
@@ -31,12 +31,16 @@ vm_species_new_hamil(struct gkyl_vm *vm_app_inp, struct gkyl_vlasov_app *app, st
   }
 
   if (vms->model_id == GKYL_MODEL_DEFAULT || vms->model_id == GKYL_MODEL_SR) {
-    // Hamiltonain is only a function of velocity space. 
-    vms->hamil_range = vms->local_vel; 
-    vms->hamil = mkarr(app->use_gpu, vms->basis_vel.num_basis, vms->local_vel.volume);
+    // Hamiltonain (moments) is only a function of velocity space.
```
Comment (JunoRavin) [inline vlasov/apps/vm_species.c]: Can we make this comment a little more clear (Hamiltonian *for* computing moments just like what you have in the header)

File: vlasov/apps/vlasov.c [hunk 2]
```diff
@@ -948,12 +952,17 @@ gkyl_vlasov_app_from_file_field(gkyl_vlasov_app *app, const char *fname)
   struct gkyl_app_restart_status rstat = header_from_file(app, fname);
 
   if (rstat.io_status == GKYL_ARRAY_RIO_SUCCESS) {
+    // Fixed-function field BCs are frozen from the initial conditions, so seed
+    // those buffers before the local solution is overwritten by restart data.
+    gkyl_vlasov_app_apply_ic_field(app, 0.0);
```
Comment (JunoRavin) [inline vlasov/apps/vlasov.c:957]: Have you tested restarts? From the look of this program flow, you're not re-scaling the EM fields with Jc at restart, but if I recall you are writing out the undensitized fields

Comment (johnson452) [reply vlasov/apps/vlasov.c:957]: Added a rescaling by Jc in commit: [352e685](https://github.com/ammarhakim/gkeyll/pull/1046/commits/352e685ba72f1a5efd02bfeb6e8f69bdc5773439)

File: vlasov/apps/vm_species.c (hunk 1 above)
Comment (johnson452) [reply vlasov/apps/vm_species.c]: Fixed.

**Comment by johnson452** [comment]
This PR supersedes PR: https://github.com/ammarhakim/gkeyll/pull/963 on the gr triads and includes all the work from that branch. This branch, dg_gr_maxwell, resolves merge conflicts with both main and kitp-hamil-vlasov. The gr_triads PR has therefore been close.

**Comment by JunoRavin** [comment]
Ready to merge
